src/codegen/encoding/mod.rs

```diff
@@ -289,16 +765,15 @@ fn calculate_array_size<T: AbiEncoding>(
             vartab,
             cfg,
         );
-        size_var
-    };
-
-    Expression::Variable(Loc::Codegen, Type::Uint(32), size_var)
+        Expression::Variable(Codegen, Uint(32), size_var)
+    }
 }
 
 /// Calculate the size of a complex array.
-/// This function indexes an array from its outer dimension to its inner one
-fn calculate_complex_array_size<T: AbiEncoding>(
-    encoder: &mut T,
+/// This function indexes an array from its outer dimension to its inner one and
+/// accounts for the encoded length size for dynamic dimensions.
+fn calculate_complex_array_size(
```

Comment:
Why is `encode_complex_array` a member of `trait AbiEncoding` while `calculate_complex_array_size` is not?
These functions are pretty similar.

Reply:
When I implemented this, I envisioned that the functions that could be common for multiple encoding schemes are the standalone ones, which received the encoder as a generic type argument. Those that should be specific to an encoding algorithm are implemented in its corresponding file.

A standalone function that receives a generic argument and a default implementation for the trait are basically the same. Now we have both methods in the file. Perhaps, it is better to stick to a single strategy to provide common methods for both encoding schemes.

Reply:
The array size calculation functions were not in the trait before I touched it (only the encode method), so I left it that way. 

I agree that the interface we have here is currently a bit weird. I discussed this already with Sean another day
- The way it is now might or might not be suitable for other encoding schemes
- Depending on their needs, we might need further improvements to this interface
- XDR (Soroban / Stellar) and ETH ABI encoding would be good candidates to analyze
- Hence, before we implement the next encoding scheme, we reshape this API if necessary.

This is way beyond the scope of the PR, so I'm gonna leave it as it is.

Reply:
> The array size calculation functions were not in the trait before I touched it (only the encode method), so I left it that way.

None of the functions you added to the trait were already there, but you added them anyways. `encode_complex_array` was not in the trait. Can you think of a case when someone needs to override `encode_comple_array` but not `calculate_complex_array_size`?

> I discussed this already with Sean another day

You should start providing either a design document or the context of your discussions in your PR, so we don't have this misunderstandings anymore.

Reply:
That makes sense. Thanks!